### DISCUSSION TITLE
Is there any plan to add Generic, Pick, Omit and other TypeScript like features?

### DISCUSSION THREAD

**Discussion by razb-viola** [OAI/OpenAPI-Specification#3406] [Q&A]
In TS if you want to $ref specific properties from a Schema, it's a pain, at least from what I understood.
For example:

```
//permission
{
  id: string;
  key: string;
  info: string;
  metadata: Metadata;
}
// role
{
  id: string;
  title: string;
  permissions: {
    id: Permissions.id;
    key: Permissions.key;
    // notice that info shouldn't be here.
  }
}
```

As you can see, there is a relation between Permission and Role. But Role shouldn't include all of the Permission properties, just part of them. Imagine if Permission or other embedded Schemas had 50 properties, this could lead to big issue.

**

Moreover, if we want to make it even more tricky, in a real life production API, we have different kinds of the same Schema, for example: when creating a new Permission we should never provide id. If someone provides id we want to deny the request. When updating a Permission, we can only update the info, but never the key. 
When returning a Permission in a GET request, we would like to return the entire Permission, but without the metadata (which is basically some system info).

Any advices here?

**Comment by handrews** [comment]
This is a JSON Schema question rather than an OpenAPI question, so you might want to raise it with them or consider proposing an extension vocabulary.

OpenAPI 3.1 made a point to be more compatible with standard JSON Schema, with fewer custom extensions, and [OpenAPI 4.0 will be more loosely coupled with content schema formats](https://www.openapis.org/blog/2023/12/06/openapi-moonwalk-2024).  So this request is out of scope for this project.

**Reply by handrews** [reply]
@razb-viola for your comments about managing variations for permissions vs http requests and which fields are allowed in which combinations, I think there are [several discussions for OpenAPI 4 / Moonwalk](https://github.com/OAI/moonwalk/discussions) that might address those concerns.

* https://github.com/OAI/moonwalk/discussions/30
* https://github.com/OAI/moonwalk/discussions/49
* https://github.com/OAI/moonwalk/discussions/13
* https://github.com/OAI/moonwalk/discussions/60

and probably others, those were the first few I saw that seemed at least kinda relevant.
